### PR TITLE
Moved `move`, `moveEmplace`, and `forward` to `core.lifetime`

### PR DESCRIPTION
**Pull request by TurkeyMan**
WIP, I still need to wrangle phobos dependencies...

### CODE REVIEW

**Comment by TurkeyMan** [comment]
Yup, is WIP. Just here so people can complain eagerly ;)

**Comment by TurkeyMan** [comment]
WTF is with Phobos!! The dependence hierarchies are all completely insane!
This is basically impossible, the dependency graph is just wild... `move` pulls the whole universe!
Check out C++'s `std::move`, it should be that simple!

**Comment by TurkeyMan** [comment]
I'm not sure what the go with this error is...
I cut and pasted the code from phobos verbatim, and this test works there :/

**Comment by wilzbach** [comment]
CI says:

```
src/core/lifetime.d(1904): Error: struct `core.lifetime.__unittest_L1934_C7.NoCopy` is not copyable because it is annotated with `@disable`
src/core/lifetime.d(1715): Error: template instance `core.lifetime.moveEmplace!(NoCopy[2])` error instantiating
src/core/lifetime.d(1709):        instantiated from here: `moveImpl!(NoCopy[2])`
src/core/lifetime.d(1667):        instantiated from here: `trustedMoveImpl!(NoCopy[2])`
src/core/lifetime.d(1948):        instantiated from here: `move!(NoCopy[2])`
src/core/lifetime.d(1904): Error: struct `core.lifetime.__unittest_L1934_C7.NoCopy` is not copyable because it is annotated with `@disable`
src/core/lifetime.d(1715): Error: template instance `core.lifetime.moveEmplace!(NoCopy[2])` error instantiating
src/core/lifetime.d(1709):        instantiated from here: `moveImpl!(NoCopy[2])`
src/core/lifetime.d(1667):        instantiated from here: `trustedMoveImpl!(NoCopy[2])`
src/core/lifetime.d(1948):        instantiated from here: `move!(NoCopy[2])`
```

Does 


```
make -f posix.mak unittest-debug
```

work for your locally?

**Comment by thewilsonator** [comment]
It fails because you haven't copied all of `moveEmplace`, missing is:
```
    else static if (isStaticArray!T)
    {
        for (size_t i = 0; i < source.length; ++i)
            move(source[i], target[i]);
    }
```

**Comment by TurkeyMan** [comment]
Oh.... shit.
I wasn't copying from phobos HEAD, I accidentally pressed "go to definition" in my editor, and it opened the version of the file from my local DMD import path from my local compiler install, which was from the last release >_<

I hope I haven't done that elsewhere!! O_O
I need to audit everything I've done!

**Comment by wilzbach** [comment]
> Is this G2G? That one test fail looks completely unrelated...

Yup. Deprecated the build for you.

**Comment by TurkeyMan** [comment]
What does that mean? Can this merge?

File: src/core/lifetime.d [hunk 1]
```diff
@@ -1292,3 +1292,666 @@ pure nothrow @safe /* @nogc */ unittest
...
+void moveEmplace(T)(ref T source, ref T target) @system
+{
+    import core.stdc.string : memcpy, memset;
+    import core.internal.traits;
+
+    // TODO: this assert pulls in half of phobos. we need to work out an alternative assert strategy.
+//    static if (!is(T == class) && hasAliasing!T) if (!__ctfe)
+//    {
+//        import std.exception : doesPointTo;
+//        assert(!doesPointTo(source, source), "Cannot move object with internal pointer.");
```
Comment (wilzbach) [inline src/core/lifetime.d]: How do we ensure we fix this before the next release?

**Comment by wilzbach** [comment]
> How do we ensure we fix this before the next release?

Okay, I ported `doesPointTo` for you to druntime. Feel free to move it to `core.internal` or wherever it belongs. I didn't port `hasAliasing` as it pulled in too much for now.

**Comment by wilzbach** [comment]
> I'm not sure your patch is right... you're missing the hasAliasing guard, which will make the assert more inclusive... false positives?

I'm aware and I wanted to port `hasAliasing` too. I just ran out of time.

> But if you wanna keep it, can you move doesPointTo to core.internal.traits? That's the appropriate dumping ground. I don't want core.lifetime to already be a dumping ground at the moment of its birth...

Yeah I know. Just wanted to quickly test whether it works on all CIs and didn't have much time.

**Comment by wilzbach** [comment]
> That's the appropriate dumping ground

So yeah I managed to port `hasAliasing` and `doesPointTo`:

```
 2 files changed, 1922 insertions(+), 6 deletions(-)
```

Though of course we can remove all these lines from Phobos and replace them with aliases.

**Comment by wilzbach** [comment]
> We can't remove them from phobos unless we put then in a public module, with doco generated and tests... then we can make phobos aliases.

Why shouldn't we be able to alias them? Even DMD's documentation generator supports exposing aliased symbols. We would just need to keep the public tests in Phobos.

**Comment by wilzbach** [comment]
> concluded that it was completely unreasonable to draw on that much code 

Not sure as we'll probably end up pulling most of this code from druntime at some point anyhow.

**Comment by thewilsonator** [comment]
Auto-merge toggled on

**Comment by wilzbach** [comment]
FYI started to remove the now duplicate implementations in Phobos:
- forward: https://github.com/dlang/phobos/pull/6847
- move{,Emplace}: https://github.com/dlang/phobos/pull/6848

But we can't move `emplace`  yet as we need a public version of `emplaceRef`. Though imho we should simply set it public as emplaceRef is used all over in Phobos (see e.g. https://github.com/dlang/phobos/search?q=emplaceRef&unscoped_q=emplaceRef).
It's also part of the proposed RCArray PR for druntime (https://github.com/dlang/druntime/pull/2348).

CC @edi33416

**Comment by andralex** [comment]
We can for now make `emplaceRef` public and docummented. @wilzbach would that work?

**Comment by wilzbach** [comment]
> We can for now make `emplaceRef` public and docummented. @wilzbach would that work?

Yup, that's basically what I proposed ;-)

**Comment by andralex** [comment]
Another possibility is to overload, where are the ambiguities?

**Pull request merged**